lib/grant.js

```diff
@@ -449,7 +452,24 @@ function saveRefreshToken (done) {
 }
 
 /**
- * Create an access token and save it with the model
+ * Pass the scope string to the model for saving
+ *
+ * @param  {Function} done
+ * @this   OAuth
+ */
+function saveScope (done) {
+  var scope = this.scope || this.req.body.scope;
+
+  var self = this;
+  this.model.saveScope(scope, self.accessToken, function (err, acceptedScope) {
+    if (err) return done(error('server_error', false, err));
```

Comment:
Yup, I felt the same pain. I had your PR ready to go but I'm holding off on submitting because I felt the need to create a new error architecture, exactly as you described.

I incorporated almost every change we discussed in this PR, with the exception of passing `scope` to the `saveAccessToken()` function since it prevents me from throwing an "invalid scope" error, as I described previously. I also added tests for all of the new functionality.

You can take a look at my commits here: https://github.com/seegno-forks/node-oauth2-server/commits/feature-scope but I'll hold off on submitting it until @thomseddon weighs in on this.